### ISSUE TITLE
shapeless.Coproduct is imported but not used

### ISSUE THREAD

**Issue by zcox** [julianpeeters/avrohugger#83]
Generated code that uses Shapeless Coproduct includes this import:

```scala
import shapeless.{:+:, CNil, Coproduct}
```

[Example](https://github.com/julianpeeters/avrohugger/blob/master/avrohugger-core/src/test/expected/standard/example/idl/AllUnionsWithShapelessCoproduct.scala#L4).

However, it seems that the generated code doesn't actually use `shapeless.Coproduct`, which causes the following compiler error (I think if you have both `-Xfatal-warnings` and `-Ywarn-unused:imports`):

```
[error] /Users/foo/bar/target/scala-2.12/src_managed/test/compiled_avro/com/foo/Bar.scala:4: Unused import
[error] import shapeless.{:+:, CNil, Coproduct}
[error]                              ^
[error] one error found
```

**Comment by zcox** [comment]
`shapeless.Coproduct` is used in the generated Scala file for default values.

```
    record RecordWithADefaultValue {
        union { int, string, boolean } isb = 0;
    }
```

```scala
case class RecordWithADefaultValue(isb: Int :+: String :+: Boolean :+: CNil = Coproduct[Int :+: String :+: Boolean :+: CNil](0))
```

It would be great if the `shapeless.Coproduct` import could just be omitted when it is not used in the generated Scala.

**Comment by david-mcgillicuddy-ovo** [comment]
Hello I started to look at what this would require: https://github.com/julianpeeters/avrohugger/compare/master...david-mcgillicuddy-ovo:master
However I seem to have somehow broken 94/283 tests doing this. Whole swathes of examples that I haven't touched get a lot of "Not a record" errors now, when those tests pass on clean master. I'd appreciate any suggestions please!

**Comment by julianpeeters** [comment]
Hey @david-mcgillicuddy-ovo, thanks for taking a look, and sorry you ran into that! I believe what is going on here is that I let slip out a bad release which broke factored schemas (schemas whose fields reference types by name instead of by defining a nested schema in one file). https://github.com/julianpeeters/avrohugger/commit/47b50979e475a04af5b23b403bd9c34577937624


That code will need to be able to handle already-defined schemas in _both_ cases
1) parse factored schemas
2) exclude true duplicate schemas from parsing



I believe I had an oversight on what was required, combined with an sbt caching issue that didn't catch this in tests. Obviously this is kind of serious issue, but I don't think I'll have time to fix it until the weekend.


As a workaround, as Zach mentions above, providing a default value will solve the unused import error (and is a good practice if you ever want to remove the field later).

**Comment by julianpeeters** [comment]
Thanks for your patience, gentlemen. This is fixed now in avrohugger verson 1.0.0-RC7

**Comment by rkorver** [comment]
Hi @julianpeeters 

When using a union of more than 2 types, using `sbt compile`, (plugin version 2.0.0),  I'm getting this "unused import" error in the parent generated class.

https://stackoverflow.com/questions/41478202/can-we-use-oneof-in-confluent-platform-schema-registry (the schema in the response works -- also with 3 types in the union.) 
 
The schema below doesn't work, even though the avro schema is/seems valid:
 
 ```
{
  "name": "userBalance",
  "type": "record",
  "fields": [
    {
      "name": "currentBalance",
      "type": {
        "type": "record",
        "name": "CurrentBalance",
        "fields": [
          {
            "name": "userBalance",
            "type": [
              {
                "type": "record",
                "name": "userCashDeposit",
                "fields": [
                  {
                    "name": "id",
                    "type": "long"
                  },
                  {
                    "name": "amount",
                    "type": "float"
                  }
                ]
              },
              {
                "type": "record",
                "name": "userCashWithdraw",
                "fields": [
                  {
                    "name": "id",
                    "type": "long"
                  },
                  {
                    "name": "amount",
                    "type": "float"
                  }
                ]
              },
              {
                "type": "record",
                "name": "userBalanceCheck",
                "fields": [
                  {
                    "name": "id",
                    "type": "long"
                  },
                  {
                    "name": "amount",
                    "type": "float"
                  }
                ]
              }
            ]
          }
        ]
      }
    }
  ]
}
```
Yielding: 
``` 
my_path/userBalance.scala:2:24: Unused import
[error] import shapeless.{:+:, CNil}
```

I'm probably missing something here... but if it's a bug, might you have a fix / workaround for this?
